In [3]:
# The first step of using numpy is to tell python to import it
import numpy as np

### 2.1 NumPy Arrays

**NumPy Array**
* An array is a data structure that stores values of same data type.
* While python lists can contain values corresponding to different data types, arrays in python can only contain values corresponding to the same data type.
* However python lists fail to deliver the performance required while computing large sets of numerical data. To address this issue we use NumPy arrays.
* We can create NumPy arrays by converting a list to an array.


In [4]:
# defining a list of different car companies or string elements
arr_str = ['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla']

# defining a list of number of cylinders in car or numerical elements
arr_num = [5, 4, 6, 7, 3]

In [5]:
temp = np.array(['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla'])
print(type(temp))
print(temp)

<class 'numpy.ndarray'>
['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']


In [6]:
# connverting the list arr_str to a NumPy array
np_arr_str = np.array(arr_str)


# connverting the list arr_num to a NumPy array
np_arr_num = np.array(arr_num)

# checking the output
print('Numpy Array (arr_str): ',np_arr_str)
print('Numpy Array (arr_num): ',np_arr_num)

Numpy Array (arr_str):  ['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']
Numpy Array (arr_num):  [5 4 6 7 3]


The resuts look similar to a list but arr_str and arr_num have been converted to NumPy arrays. Let's check the data type to confirm this.

In [7]:
# printing the data type of lists
print('Data type of arr_str: ',type(arr_str))
print('Data type of arr_num: ',type(arr_num))

# printing the data type after conversion of lists to array
print('Data type of np_arr_str: ',type(np_arr_str))
print('Data type of np_arr_num: ',type(np_arr_num))

Data type of arr_str:  <class 'list'>
Data type of arr_num:  <class 'list'>
Data type of np_arr_str:  <class 'numpy.ndarray'>
Data type of np_arr_num:  <class 'numpy.ndarray'>


* The above output confirms that both the lists were successfully converted to arrays

**NumPy Matrix**

* A matrix is a two-dimensional data structure where elements are arranged into rows and columns.
* A matrix can be created by using list of lists

In [8]:
# let's say we have information of different number of cylinders in a car and we want to display them in a matrix format
matrix = np.array([[1,2,1],[4,5,9],[1,8,9]])
print(matrix)

[[1 2 1]
 [4 5 9]
 [1 8 9]]


In [9]:
print('Data type of matrix: ',type(matrix))

Data type of matrix:  <class 'numpy.ndarray'>


* We see that all the NumPy objects have data type as ndarray

### 2.2 NumPy Functions

**There are different ways to create NumPy arrays using the functions available in NumPy library**

**Using np.arange() function**
* The np.arange() function returns an array with evenly spaced elements as per the interval. The interval mentioned is half-opened i.e. start is included but stop is excluded.
* It has the following paramaters:
  * start : start of interval range. By default start = 0
  * stop  : end of interval range
  * step  : step size of interval. By default step size = 1

In [10]:
list(range(0, 10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [11]:
arr2  = np.arange(start = 0, stop = 10) # 10 will be excluded from the output
print(arr2)

# or

arr2  = np.arange(0,10)
print(arr2)

[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3 4 5 6 7 8 9]


In [12]:
temp = np.arange(0, 101, 2)
print(temp)

[  0   2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34
  36  38  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70
  72  74  76  78  80  82  84  86  88  90  92  94  96  98 100]


In [13]:
# adding a step size of 5 to create an array
arr3  = np.arange(start = 0, stop = 21, step = 5)
print(arr3)

[ 0  5 10 15 20]


**Using np.linspace() function**
* The np.linspace() function returns numbers which are evenly distributed with respect to interval. Here the start and stop both are included.            
*It has the following parameters:              
 * start: start of interval range. By default start = 0
 * stop: end of interval range
 * num : No. of samples to generate. By default num = 50

In [14]:
matrix2 = np.linspace(0,5) # by default 50 evenly spaced values will be generated between 0 and 5
matrix2

array([0.        , 0.10204082, 0.20408163, 0.30612245, 0.40816327,
       0.51020408, 0.6122449 , 0.71428571, 0.81632653, 0.91836735,
       1.02040816, 1.12244898, 1.2244898 , 1.32653061, 1.42857143,
       1.53061224, 1.63265306, 1.73469388, 1.83673469, 1.93877551,
       2.04081633, 2.14285714, 2.24489796, 2.34693878, 2.44897959,
       2.55102041, 2.65306122, 2.75510204, 2.85714286, 2.95918367,
       3.06122449, 3.16326531, 3.26530612, 3.36734694, 3.46938776,
       3.57142857, 3.67346939, 3.7755102 , 3.87755102, 3.97959184,
       4.08163265, 4.18367347, 4.28571429, 4.3877551 , 4.48979592,
       4.59183673, 4.69387755, 4.79591837, 4.89795918, 5.        ])

**How are these values getting generated?**

The step size or the difference between each element will be decided by the following formula:

**(stop - start) / (total elements - 1)**

So, in this case:
(5 - 0) / 49 = 0.10204082

The first value will be 0.10204082, the second value will be 0.10204082 + 0.10204082, the third value will be 0.10204082 + 0.10204082 +0.10204082, and so on.

In [15]:
(10-0)/49

0.20408163265306123

In [16]:
np.linspace(0, 10)

array([ 0.        ,  0.20408163,  0.40816327,  0.6122449 ,  0.81632653,
        1.02040816,  1.2244898 ,  1.42857143,  1.63265306,  1.83673469,
        2.04081633,  2.24489796,  2.44897959,  2.65306122,  2.85714286,
        3.06122449,  3.26530612,  3.46938776,  3.67346939,  3.87755102,
        4.08163265,  4.28571429,  4.48979592,  4.69387755,  4.89795918,
        5.10204082,  5.30612245,  5.51020408,  5.71428571,  5.91836735,
        6.12244898,  6.32653061,  6.53061224,  6.73469388,  6.93877551,
        7.14285714,  7.34693878,  7.55102041,  7.75510204,  7.95918367,
        8.16326531,  8.36734694,  8.57142857,  8.7755102 ,  8.97959184,
        9.18367347,  9.3877551 ,  9.59183673,  9.79591837, 10.        ])

In [17]:
(20-10)/9

1.1111111111111112

In [18]:
# generating 10 evenly spaced values between 10 and 20
matrix3 = np.linspace(10,20,10)
matrix3

array([10.        , 11.11111111, 12.22222222, 13.33333333, 14.44444444,
       15.55555556, 16.66666667, 17.77777778, 18.88888889, 20.        ])

**Similarly we can create matrices using the functions available in NumPy library**

**Using np.zeros()**

* The np.zeros() is a function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix filled with zeros of the given shape.
* It has the following parameters:    
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [19]:
matrix4 = np.zeros([3,5])
matrix4

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [20]:
temp = np.zeros([3,5], int)
temp

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

**Using np.ones()**

* The np.ones() is another function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix of given shape and type, filled with ones.
* It has the following parameters:  
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [21]:
matrix5 = np.ones([3,5])
matrix5

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

**Using np.eye()**
* The np.eye() is a function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix with ones on the diagonal and zeros elsewhere.
* It has the following parameters:
  * n: Number of rows and columns in the output matrix
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [22]:
matrix6 = np.eye(5)
matrix6

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [23]:
matrix6 = np.eye(5, dtype = int)
matrix6

array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1]])

**We can also convert a one dimension array to a matrix. This can be done by using the np.reshape() function.**

* The shape of an array basically tells the number of elements and dimensions of the array. Reshaping a Numpy array simply means changing the shape of the given array.
* By reshaping an array we can add or remove dimensions or change number of elements in each dimension.
* In order to reshape a NumPy array, we use the reshape method with the given array.
* **Syntax:** array.reshape(shape)
  * shape: a tuple given as input, the values in tuple will be the new shape of the array.

In [24]:
# defining an array with values 0 to 9
arr4 = np.arange(0,10)
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [25]:
# reshaping the array arr4 to a 2 x 5 matrix
arr4_reshaped = arr4.reshape((2,5))
arr4_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [26]:
# reshaping the array arr4 to a 2 x 5 matrix
arr4_reshaped = arr4.reshape([2,5])
arr4_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [27]:
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [28]:
# reshaping the array arr4 to a 2 x 6 matrix
# arr4.reshape((2,6))

* This did not work because we have 10 elements which we are trying to fit in a 2 X 6 shape which will require 12 elements.

**NumPy can also perform a large number of different mathematical operations and it provides different functions to do so.**

NumPy provides:
1. Trigonometric functions
2. Exponents and Logarithmic functions
3. Functions for arithmetic operations between arrays and matrices

**Trigonometric functions**

In [29]:
print('Sine Function:',np.sin(4))
print('Cosine Function:',np.cos(4))
print('Tan Function',np.tan(4))

Sine Function: -0.7568024953079282
Cosine Function: -0.6536436208636119
Tan Function 1.1578212823495775


**Exponents and Logarithmic functions**

* Exponents

In [30]:
np.exp(2)

7.38905609893065

In [31]:
arr5 = np.array([2,4,6])
np.exp(arr5)

array([  7.3890561 ,  54.59815003, 403.42879349])

* Logarithms

In [32]:
# by default NumPy takes the base of log as e
np.log(2)

0.6931471805599453

In [33]:
np.log(arr5)

array([0.69314718, 1.38629436, 1.79175947])

In [34]:
## log with base 10
np.log10(8)

0.9030899869919435

In [35]:
## log with base 10
np.log10(arr5)

array([0.30103   , 0.60205999, 0.77815125])

**Arithmetic Operations on arrays**

In [36]:
# arithmetic on lists

l1 = [1,2,3]
l2 = [4,5,6]
print(l1+l2)
print(l1+l2)
# this does not behave as you would expect!


[1, 2, 3, 4, 5, 6]
[1, 2, 3, 4, 5, 6]


In [37]:
# we can +-*/ arrays together

# defining two arrays
arr7 = np.arange(1,6)
print('arr7:', arr7)

arr8 = np.arange(3,8)
print('arr8:', arr8)

arr7: [1 2 3 4 5]
arr8: [3 4 5 6 7]


In [38]:
5**7

78125

In [39]:
print('Addition: ',arr7+arr8)
print('Subtraction: ',arr8-arr7)
print('Multiplication:' , arr7*arr8)
print('Division:', arr7/arr8)
print('Inverse:', 1/arr7)
print('Powers:', arr7**arr8) # in python, powers are achieved using **, NOT ^!!! ^ does something completely different!

Addition:  [ 4  6  8 10 12]
Subtraction:  [2 2 2 2 2]
Multiplication: [ 3  8 15 24 35]
Division: [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse: [1.         0.5        0.33333333 0.25       0.2       ]
Powers: [    1    16   243  4096 78125]


**Operations on Matrices**

In [40]:
matrix7 = np.arange(1,10).reshape(3,3)
print(matrix7)

matrix8 = np.eye(3)
print(matrix8)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [41]:
print('Addition: \n', matrix7+matrix8)
print('Subtraction: \n ', matrix7-matrix8)
print('Multiplication: \n', matrix7*matrix8)
print('Division: \n', matrix7/matrix8)
print('Power: \n', matrix7**matrix8)

Addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
  [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]
Power: 
 [[1. 1. 1.]
 [1. 5. 1.]
 [1. 1. 9.]]


<ipython-input-41-5dddfc2aa5ac>:4: RuntimeWarning: divide by zero encountered in divide
  print('Division: \n', matrix7/matrix8)


* RuntimeWarning: Errors which occur during program execution(run-time) after successful compilation are called run-time errors.
* One of the most common run-time error is division by zero also known as Division error.
* Due to division by zero error, we are getting inf (infinity) values because 1/0 is not a defined operation.

**Linear algebra matrix multiplication**

In [42]:
matrix9 = np.arange(1,10).reshape(3,3)
print('First Matrix: \n',matrix9)

matrix10 = np.arange(11,20).reshape(3,3)
print('Second Matrix: \n',matrix10)
print('')
# taking linear algebra matrix multiplication (some may have heard this called the dot product)
print('Multiplication: \n', matrix9 @ matrix10)

First Matrix: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Second Matrix: 
 [[11 12 13]
 [14 15 16]
 [17 18 19]]

Multiplication: 
 [[ 90  96 102]
 [216 231 246]
 [342 366 390]]


In [43]:
1*11 + 2*14 + 3*17, 4*11 + 5*14 + 6*17, 7*11 + 8*14 + 9*17

(90, 216, 342)

**Transpose of a matrix**

In [44]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [45]:
# taking transpose of matrix
np.transpose(matrix9)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [46]:
# another way of taking a transpose
matrix9.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

**Function to find minimum and maximum values**

In [47]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [48]:
print('Minimum value: ',np.min(matrix9))

Minimum value:  1


In [49]:
print('Maximum value: ',np.max(matrix9))

Maximum value:  9


**Function to generate random samples**

**Using np.random.rand function**

* The np.random.rand returns a random NumPy array whose element(s) are drawn randomly from the uniform distribution over [0,1). (including 0 but excluding 1).
* **Syntax** - np.random.rand(d0,d1)
  * d0,d1 – It represents the dimension of the required array given as int, where d1 is optional.

In [50]:
# Generating random values in an array
rand_mat = np.random.rand(5)
print(rand_mat)

[0.26153852 0.94411494 0.11457211 0.14299629 0.41415172]


In [51]:
# * Generating random values in a matrix
rand_mat = np.random.rand(5,5) # uniform random variable
print(rand_mat)

[[0.0949829  0.32649701 0.46058535 0.43367874 0.02263151]
 [0.79204683 0.4162047  0.01836867 0.94222017 0.42402526]
 [0.65339529 0.66001772 0.80357184 0.15582053 0.25109114]
 [0.47767558 0.95879968 0.68151977 0.75702763 0.03973675]
 [0.9526478  0.74382948 0.9685702  0.06145104 0.05564851]]


**Using np.random.randn function**

* The np.random.randn returns a random numpy array whose sample(s) are drawn randomly from the standard normal distribution (Mean as 0 and standard deviation as 1)

* **Syntax** - np.random.randn(d0,d1)
  * d0,d1 – It represents the dimension of the output, where d1 is optional.

In [52]:
# Generating random values in an array
rand_mat2 = np.random.randn(5)
print(rand_mat2)

[-0.76473396  0.88922679 -1.29930998 -0.98958898 -0.13559032]


In [53]:
# Generating random values in a matrix
rand_mat2 = np.random.randn(500,500)
print(rand_mat2)

[[ 0.96673219 -0.82999961 -1.62126464 ...  0.69334251 -0.16184206
  -0.40109373]
 [ 1.3787694   1.27360766 -0.45498967 ... -0.43236047  2.22331137
  -0.03265891]
 [-0.8734371   0.85524717 -0.31093042 ... -1.06717349  0.95176071
  -1.53301051]
 ...
 [ 0.98303731 -0.42963306 -0.67693228 ...  1.02461235 -0.82962061
  -0.39592216]
 [-0.18585211  2.79367983  0.58093852 ...  1.30691668  0.54944001
   0.90045903]
 [ 0.93860408 -1.53856451 -0.96434734 ...  0.17468543 -0.68693177
   1.24227064]]


In [54]:
# Let's check the mean and standard deviation of rand_mat2
print('Mean:',np.mean(rand_mat2))
print('Standard Deviation:',np.std(rand_mat2))

Mean: -0.002133834059600346
Standard Deviation: 0.9975739328644998


*  We observe that the mean is very close to 0 and standard deviation is very close to 1.

**Using np.random.randint function**

* The np.random.randint returns a random numpy array whose element(s) are drawn randomly from low (inclusive) to the high (exclusive) range.

* **Syntax** - np.random.randint(low, high, size)

  * low – It represents the lowest inclusive bound of the distribution from where the sample can be drawn.
  * high – It represents the upper exclusive bound of the distribution from where the sample can be drawn.
  * size – It represents the shape of the output.

In [55]:
# Generating random values in an array
rand_mat3 = np.random.randint(1,5,10)
print(rand_mat3)

[3 2 1 1 4 1 4 2 1 2]


In [56]:
# Generating random values in a matrix
rand_mat3 = np.random.randint(1,10,[5,5])
print(rand_mat3)

[[8 7 5 7 8]
 [5 4 5 2 7]
 [5 8 9 1 4]
 [2 6 3 7 5]
 [5 1 6 6 5]]


### 2.3 Accessing the entries of a Numpy Array

In [58]:
l = [3, 4, 5, 6]
l[0]

3

In [57]:
# let's generate an array with 10 random values
rand_arr = np.random.randn(10)
print(rand_arr)

[ 0.70186512  0.95152324  0.06111668 -0.47103435  0.72445532 -0.24676129
 -0.29135383 -0.90674941 -0.82510543  0.98856299]


* Accessing one element from an array

In [60]:
# accessing the 6 th entry of rand_arr
print(rand_arr[7])

-0.9067494146897354


* Accessing multiple elements from an array

In [65]:
# we can access multiple entries at once using
print(rand_arr[4:9])

[ 0.72445532 -0.24676129 -0.29135383 -0.90674941 -0.82510543]


In [66]:
# we can also access multiple non-consecutive entries using np.arange
print('Index of values to access: ',np.arange(3,10,3))
print(rand_arr[np.arange(3,10,3)])

Index of values to access:  [3 6 9]
[-0.47103435 -0.29135383  0.98856299]


**Accessing arrays using logical operations**

In [67]:
print(rand_arr)

[ 0.70186512  0.95152324  0.06111668 -0.47103435  0.72445532 -0.24676129
 -0.29135383 -0.90674941 -0.82510543  0.98856299]


In [68]:
rand_arr>0

array([ True,  True,  True, False,  True, False, False, False, False,
        True])

In [69]:
# accessing all the values of rand_arr which are greater than 0
print('Values greater than 0: ',rand_arr[rand_arr>0])

# accessing all the values of rand_arr which are less than 0
print('Values less than 0: ',rand_arr[rand_arr<0])

Values greater than 0:  [0.70186512 0.95152324 0.06111668 0.72445532 0.98856299]
Values less than 0:  [-0.47103435 -0.24676129 -0.29135383 -0.90674941 -0.82510543]


**Accessing the entries of a Matrix**

In [74]:
print(rand_mat3)

[[8 7 5 7 8]
 [5 4 0 2 7]
 [5 8 9 1 4]
 [2 6 3 7 5]
 [5 1 6 6 5]]


In [73]:
rand_mat3[1][2] = 0
print(rand_mat3)

[[8 7 5 7 8]
 [5 4 0 2 7]
 [5 8 9 1 4]
 [2 6 3 7 5]
 [5 1 6 6 5]]


In [77]:
rand_mat3[1,2]

0

In [75]:
# let's generate an array with 10 random values
rand_mat = np.random.randn(5,5)
print(rand_mat)

[[-1.52876741 -0.45958668  0.60535048 -1.36832623  0.01648405]
 [ 0.24342868 -0.15984966 -0.07557094 -0.25823382  0.19053627]
 [ 1.45888059  1.00153229  0.01682585 -0.05255299  0.38453373]
 [ 1.28142806  0.13172767  0.9916329   0.16385619  0.30219937]
 [-1.56126712  0.13561618 -0.2696863  -0.00860327  0.53868966]]


In [76]:
# acessing the second row of the rand_mat
rand_mat[1]

array([ 0.24342868, -0.15984966, -0.07557094, -0.25823382,  0.19053627])

In [78]:
# acessing third element of the second row
print(rand_mat[1][2])

#or

print(rand_mat[1,2])

-0.07557094386643451
-0.07557094386643451


In [79]:
# accessing first two rows with second and third column
print(rand_mat[0:2,1:3])

[[-0.45958668  0.60535048]
 [-0.15984966 -0.07557094]]


**Accessing matrices using logical operations**

In [80]:
print(rand_mat)

[[-1.52876741 -0.45958668  0.60535048 -1.36832623  0.01648405]
 [ 0.24342868 -0.15984966 -0.07557094 -0.25823382  0.19053627]
 [ 1.45888059  1.00153229  0.01682585 -0.05255299  0.38453373]
 [ 1.28142806  0.13172767  0.9916329   0.16385619  0.30219937]
 [-1.56126712  0.13561618 -0.2696863  -0.00860327  0.53868966]]


In [81]:
# accessing all the values of rand_mat which are greater than 0
print('Values greater than 0: \n ',rand_mat[rand_mat>0])

# accessing all the values of rand_mat which are less than 0
print('Values less than 0: \n',rand_mat[rand_mat<0])

Values greater than 0: 
  [0.60535048 0.01648405 0.24342868 0.19053627 1.45888059 1.00153229
 0.01682585 0.38453373 1.28142806 0.13172767 0.9916329  0.16385619
 0.30219937 0.13561618 0.53868966]
Values less than 0: 
 [-1.52876741 -0.45958668 -1.36832623 -0.15984966 -0.07557094 -0.25823382
 -0.05255299 -1.56126712 -0.2696863  -0.00860327]


**Modifying the entries of an Array**

In [83]:
print(rand_arr)

[ 0.70186512  0.95152324  0.06111668 -0.47103435  0.72445532 -0.24676129
 -0.29135383 -0.90674941 -0.82510543  0.98856299]


In [84]:
# let's change some values in an array!
# changing the values of index value 3 and index value 4 to 5
rand_arr[3:5] = 5
print(rand_arr)

[ 0.70186512  0.95152324  0.06111668  5.          5.         -0.24676129
 -0.29135383 -0.90674941 -0.82510543  0.98856299]


In [85]:
# changing the values of index value 0 and index value 1 to 2 and 3 respectively
rand_arr[0:2] = [2,3]
print(rand_arr)

[ 2.          3.          0.06111668  5.          5.         -0.24676129
 -0.29135383 -0.90674941 -0.82510543  0.98856299]


In [86]:
# modify entries using logical references
rand_arr[rand_arr>0] = 65
rand_arr

array([65.        , 65.        , 65.        , 65.        , 65.        ,
       -0.24676129, -0.29135383, -0.90674941, -0.82510543, 65.        ])

**Modifying the entries of a Matrix**

In [87]:
print(rand_mat3)

[[8 7 5 7 8]
 [5 4 0 2 7]
 [5 8 9 1 4]
 [2 6 3 7 5]
 [5 1 6 6 5]]


In [88]:
# changing the values of the 4th and 5th element of the second and third rows of the matrix to 0
print('Matrix before modification: \n',rand_mat3)
rand_mat3[1:3,3:5] = 0
print('Matrix after modification: \n',rand_mat3)

Matrix before modification: 
 [[8 7 5 7 8]
 [5 4 0 2 7]
 [5 8 9 1 4]
 [2 6 3 7 5]
 [5 1 6 6 5]]
Matrix after modification: 
 [[8 7 5 7 8]
 [5 4 0 0 0]
 [5 8 9 0 0]
 [2 6 3 7 5]
 [5 1 6 6 5]]


In [89]:
# extracting the first 2 rows and first 3 columns from the matrix
sub_mat = rand_mat[0:2,0:3]
print(sub_mat)

[[-1.52876741 -0.45958668  0.60535048]
 [ 0.24342868 -0.15984966 -0.07557094]]


In [90]:
# changing all the values of the extracted matrix to 3
sub_mat[:] = 3
print(sub_mat)

[[3. 3. 3.]
 [3. 3. 3.]]


In [91]:
# what happened to rand_mat when we change sub_mat?
rand_mat

array([[ 3.        ,  3.        ,  3.        , -1.36832623,  0.01648405],
       [ 3.        ,  3.        ,  3.        , -0.25823382,  0.19053627],
       [ 1.45888059,  1.00153229,  0.01682585, -0.05255299,  0.38453373],
       [ 1.28142806,  0.13172767,  0.9916329 ,  0.16385619,  0.30219937],
       [-1.56126712,  0.13561618, -0.2696863 , -0.00860327,  0.53868966]])

In [92]:
# to prevent this behavior we need to use the .copy() method when we assign sub_mat
# this behavior is the source of MANY errors for early python users!!!

rand_mat = np.random.randn(5,5)
print(rand_mat)
sub_mat = rand_mat[0:2,0:3].copy()
sub_mat[:] = 3
print(sub_mat)
print(rand_mat)

[[-0.57044503 -1.21225387 -1.18195596  0.53965949  1.33288967]
 [-1.10764228 -0.34737116  1.29753493 -0.87575086  1.37815037]
 [-0.12072436  1.3654778  -0.54253916  0.64478572  0.91103185]
 [ 0.59754848  1.4517412   0.37958865  0.00774829 -0.30942403]
 [ 0.6100945  -0.58287628  1.1997162   0.14765114 -0.07647536]]
[[3. 3. 3.]
 [3. 3. 3.]]
[[-0.57044503 -1.21225387 -1.18195596  0.53965949  1.33288967]
 [-1.10764228 -0.34737116  1.29753493 -0.87575086  1.37815037]
 [-0.12072436  1.3654778  -0.54253916  0.64478572  0.91103185]
 [ 0.59754848  1.4517412   0.37958865  0.00774829 -0.30942403]
 [ 0.6100945  -0.58287628  1.1997162   0.14765114 -0.07647536]]


### 2.4 Saving and Loading a NumPy array

**Let's save some NumPy objects on the disk for use later!**

In [93]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [98]:
!dir

drive  sample_data


In [99]:
# creating a random matrices
randint_matrix1 = np.random.randint(1,10,10).reshape(2,5)
print(randint_matrix1)
print('')
randint_matrix2 = np.random.randint(10,20,10).reshape(2,5)
print(randint_matrix2)

[[9 4 4 4 3]
 [9 4 1 2 4]]

[[11 10 10 18 12]
 [12 14 14 18 11]]


**Using np.save() function**

In [101]:
np.save('/content/drive/MyDrive/Colab Notebooks/Week 1/file-1',randint_matrix1)

**Using np.savez() function**

In [102]:
np.savez('/content/drive/MyDrive/Colab Notebooks/Week 1/files-1',randint_matrix1=randint_matrix1,randint_matrix2=randint_matrix2)

* The files will be saved in the directory where the Jupyter Notebook is located.
* With np.save() function, we can save an array/matrix to a NumPy .npy format.
* np.savez() function has an advantage over np.save() function because with np.savez(), we can store several arrays/matrices into a single file in uncompressed .npz format.

In [103]:
# now let's load it
loaded_arr = np.load('/content/drive/MyDrive/Colab Notebooks/Week 1/file-1.npy')
loaded_multi = np.load('/content/drive/MyDrive/Colab Notebooks/Week 1/files-1.npz')

print(loaded_arr)
print('')
print(loaded_multi)

[[9 4 4 4 3]
 [9 4 1 2 4]]

NpzFile '/content/drive/MyDrive/Colab Notebooks/Week 1/files-1.npz' with keys: randint_matrix1, randint_matrix2


* We see that .npy file has been loaded but the .npz file is returning a memory location.
* Let's see how to load the values stored in .npz file.

In [104]:
print('1st Matrix: \n',loaded_multi['randint_matrix1'])
print('2nd Matrix: \n',loaded_multi['randint_matrix2'])

new_matrix  = loaded_multi['randint_matrix1']
print('New Matrix: \n',new_matrix)

1st Matrix: 
 [[9 4 4 4 3]
 [9 4 1 2 4]]
2nd Matrix: 
 [[11 10 10 18 12]
 [12 14 14 18 11]]
New Matrix: 
 [[9 4 4 4 3]
 [9 4 1 2 4]]


In [105]:
# we can also save/load text files...but only single variables
np.savetxt('/content/drive/MyDrive/Colab Notebooks/Week 1/text_file_name.txt',randint_matrix1,delimiter=',')
rand_mat_txt = np.loadtxt('/content/drive/MyDrive/Colab Notebooks/Week 1/text_file_name.txt',delimiter=',')
print(randint_matrix1)
print('')
print(rand_mat_txt)

[[9 4 4 4 3]
 [9 4 1 2 4]]

[[9. 4. 4. 4. 3.]
 [9. 4. 1. 2. 4.]]


In [124]:
rand_mat_txt_arr = np.array(rand_mat_txt, dtype=int)

In [125]:
rand_mat_txt_arr

array([[9, 4, 4, 4, 3],
       [9, 4, 1, 2, 4]])